File: native_library/apple/FirebaseANE/VisionBarcodeANE/Extensions/FreVisionBarcode.swift
/*
 *  Copyright 2018 Tua Rua Ltd.
 *
 *  Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 *
 *  http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 */

import Foundation
import FreSwift
import FirebaseMLVision

public extension VisionBarcode {
    func toFREObject() -> FREObject? {
        guard let ret = FreObjectSwift(className: "com.tuarua.firebase.vision.Barcode") else { return nil }
        ret.frame = frame
        ret["rawValue"] = rawValue?.toFREObject()
        ret.displayValue = displayValue
        ret.format = format.rawValue
        ret.cornerPoints = cornerPoints
        ret.valueType = valueType.rawValue
        ret.email = email
        ret.phone = phone
        ret.sms = sms
        ret.url = url
        ret.wifi = wifi
        ret.geoPoint = geoPoint
        ret.contactInfo = contactInfo
        ret.driverLicense = driverLicense
        ret.calendarEvent = calendarEvent
        return ret.rawValue
    }
}

public extension Array where Element == VisionBarcode {
    func toFREObject() -> FREObject? {
        guard let ret = FREArray(className: "com.tuarua.firebase.vision.Barcode",
                                 length: self.count, fixed: true) else { return nil }
        var index: UInt = 0
        for element in self {
            ret[index] = element.toFREObject()
            index+=1
        }
        return ret.rawValue
    }
}

File: native_library/android/FirebaseANE/VisionBarcode/src/main/java/com/tuarua/firebase/ml/vision/barcode/extensions/FreVisionBarcode.kt
/*
 * Copyright 2019 Tua Rua Ltd.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.tuarua.firebase.ml.vision.barcode.extensions

import com.adobe.fre.FREArray
import com.adobe.fre.FREObject
import com.google.firebase.ml.vision.barcode.FirebaseVisionBarcode
import com.tuarua.firebase.ml.vision.common.extensions.FREArray
import com.tuarua.frekotlin.*
import com.tuarua.frekotlin.geom.set

fun FirebaseVisionBarcode.toFREObject(): FREObject? {
    val ret = FREObject("com.tuarua.firebase.ml.vision.barcode.Barcode")
    ret["frame"] = boundingBox
    ret["rawValue"] = rawValue
    ret["displayValue"] = displayValue
    ret["format"] = format
    ret["cornerPoints"] = FREArray(cornerPoints)
    ret["valueType"] = valueType
    ret["email"] = email
    ret["phone"] = phone
    ret["sms"] = sms
    ret["url"] = url
    ret["wifi"] = wifi
    ret["geoPoint"] = geoPoint
    ret["contactInfo"] = contactInfo
    ret["driverLicense"] = driverLicense
    ret["calendarEvent"] = calendarEvent
    return ret
}

fun List<FirebaseVisionBarcode>.toFREObject(): FREArray? {
    return FREArray("com.tuarua.firebase.ml.vision.barcode.Barcode",
            size, true, this.map { it.toFREObject() })
}


